Improve UIX for milestones and label-filter-bar

Echomo-Xinyu opened the pull request:
### Summary:
Fixes #96 
Fixes #98 
Fixes #103 

### Changes Made:

- Update Label Filter Bar color (red to white) and text weight (unbold) to reduce contrast compared with others
- Add default selected values for milestones (all selected)
- Add default milestone for issues/PRs without a milestone
- Add sorting rule for milestones (sort by titles)

Before:

<img width="1234" alt="image" src="https://user-images.githubusercontent.com/25452497/226092467-ed6323ba-c7d7-4d8b-b2cf-35596a56ae82.png">

After:

<img width="1308" alt="image" src="https://user-images.githubusercontent.com/25452497/226253432-59e8adcf-2fbc-48d4-9ab7-f488ede54856.png">

<img width="1320" alt="Screenshot 2023-03-20 at 10 12 40 PM" src="https://user-images.githubusercontent.com/25452497/226366879-4cf77976-fe8a-4051-869c-f74f6da83822.png">


### Commit Message:

```
Let's update the UIX for milestone and label filter bar:

- the Label Filter Bar has been updated with its background colour and font-weight
- the issues/PRs without milestones have been added a default milestone for filter purpose
- the default milestone values for the filter have been updated to avoid confusion
- the milestones displayed in the filter bar will be sorted automatically to aid searching by users
```

Echomo-Xinyu commented:
Hi Prof @damithc, this is the change you asked for last time during the meeting. Due to the nature of the label filter bar (the button that gets modified) --  supporting more complicated actions compared with the other dropdown filter bar, it is very hard to get the label filter bar to have a uniform style as the other dropdown filter bar (but still possible). I have removed its colour to reduce the contrast. Do you think it is sufficient for now?

damithc commented:
@Echomo-Xinyu I see. Does the text need to be bold? Can we show a downward arrow, similar to other filters?

On a separate note, I think we should show '... selected | ... hidden' for milestones as well. In fact it can be `.../... selected` e.g., `4/5 selected`.

e.g,,

```
  Milstones         Labels
3/4 selected      8/8selected
```

Echomo-Xinyu commented:
> @Echomo-Xinyu I see. Does the text need to be bold? Can we show a downward arrow, similar to other filters?
> 
> On a separate note, I think we should show '... selected | ... hidden' for milestones as well. In fact it can be `.../... selected` e.g., `4/5 selected`.
> 
> e.g,,
> 
> ```
>   Milstones         Labels
> 3/4 selected      8/8selected
> ```

Sure I can work on this direction.

Echomo-Xinyu commented:
> On a separate note, I think we should show '... selected | ... hidden' for milestones as well. In fact it can be `.../... selected` e.g., `4/5 selected`.
> 
> e.g,,
> 
> ```
>   Milstones         Labels
> 3/4 selected      8/8selected
> ```

Hi Prof @damithc, may I clarify this part?

Right now the design is that: when no milestone is selected, it will appear as this
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/25452497/226107677-42ede2ae-d3f4-4866-9b2a-c16770b142ca.png">
And when we select certain milestones, it will display the selected milestone as here:
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/25452497/226107820-da0baa5a-0812-496c-8102-703c4b59c762.png">

When many milestones are chosen, the milestones selected are only displayed partially:
<img width="1512" alt="image" src="https://user-images.githubusercontent.com/25452497/226107917-af504531-008c-4054-8474-6611dab20ab1.png">

I personally feel that the current design suits well for CS2103 team project since we typically have 4 - 6 milestones in total (by default all teams should have v1.1, v1.2, v1.3 and v1.4 with some having v1.2b and v1.3b) and we usually either look at overall or simply one or two milestones. In such a use case, I feel it is better to have the milestone displayed clearly rather than showing the number of milestones selected. What do you think?

damithc commented:
> I personally feel that the current design suits well for CS2103 team project since we typically have 4 - 6 milestones in total (by default all teams should have v1.1, v1.2, v1.3 and v1.4 with some having v1.2b and v1.3b) and we usually either look at overall or simply one or two milestones. In such a use case, I feel it is better to have the milestone displayed clearly rather than showing the number of milestones selected. What do you think?

@Echomo-Xinyu this is fine too, but there is no reason to change behavior before/after selection. When the page is loaded initially, all open milestones should be selected by default and hence the appearance should be as follows from the beginning:
![image](https://user-images.githubusercontent.com/1673303/226112027-b1cc6350-522b-4136-963a-203d0a0e30cc.png)

There is no reason for it to appear as follows at any point:
![image](https://user-images.githubusercontent.com/1673303/226112101-a298cadc-a861-4fb1-bee3-141e703274e6.png)

damithc commented:
When no milestone is selected, it can be something like
```
Milestones
   -
```
or
```
Milestones
   None
``` 
The point is, the `Milestone` label should stay in the same place/size all the time. There is no good reason to for it to change location at various times.

Echomo-Xinyu commented:
After some preliminary investigation, I think the rationale for the current milestone design is to make use of the [multi-select in material angular](https://material.angular.io/components/select/overview#multiple-selection) which will display the field title in the options box when there is no option selected.

I think this is the simplest way to handle the multi-select issue here and also makes use of the default option by angular material. But of course, we can also do it in a way similar to the label filter bar. Which way would you prefer then? @damithc

damithc commented:
> I think this is the simplest way to handle the multi-select issue here and also makes use of the default option by angular material. But of course, we can also do it in a way similar to the label filter bar. Which way would you prefer then? @damithc

In that case we can go with the default behavior. Nevertheless, all open milestones should be selected by default, which means it will look like this from the start:
![image](https://user-images.githubusercontent.com/1673303/226112027-b1cc6350-522b-4136-963a-203d0a0e30cc.png)

Echomo-Xinyu commented:
Noted! Let me try to research on how to set default values for the multi-select case.

Echomo-Xinyu commented:
Hi Prof @damithc, I have added a default milestone with the title "without milestone" for all the issues/PRs without a milestone and added support to filter them through the dropdown filter. Do you have any more suggestions to further polish the feature or regarding the naming convention?

damithc commented:
> Hi Prof @damithc, I have added a default milestone with the title "without milestone" for all the issues/PRs without a milestone and added support to filter them through the dropdown filter. Do you have any more suggestions to further polish the feature or regarding the naming convention?

@Echomo-Xinyu can give a screenshot?

Echomo-Xinyu commented:
> @Echomo-Xinyu Looks good. Perhaps use `Without a milestone` or `Not in a milestone`?

Noted! I adopt the "Without a milestone" here.

Eclipse-Dominator commented on src/app/core/models/issue.model.ts:
```diff
@@ -123,7 +123,8 @@ export class Issue {
     this.description = Issue.updateDescription(this.hiddenDataInDescription.originalStringWithoutHiddenData);
-    this.milestone = githubIssue.milestone ? new Milestone(githubIssue.milestone) : null;
+    // githubIssue without milestone will be set to default milestone
+    this.milestone = githubIssue.milestone ? new Milestone(githubIssue.milestone) : Milestone.DefaultMilestone;
```
With this change I think we are safe the assume that the milestone property will always be non-null/defined. Let's remove the optional parameter in line 30 to reflect the changes.
```milestone?: Milestone;``` -> ```milestone: Milestone```

Eclipse-Dominator commented on src/app/issues-viewer/card-view/card-view.component.html:
```diff
@@ -29,8 +29,8 @@
-              <span *ngIf="issue.milestone" class="octicon-milestone" octicon="milestone" color="grey" size="8"></span>
-              {{ issue.milestone?.title }}
+              <span *ngIf="issue.milestone.state" class="octicon-milestone" octicon="milestone" color="grey" size="8"></span>
+              {{ issue.milestone?.title !== 'Without a milestone' ? issue.milestone?.title : '' }}
```
I think checking with `issue.milestone.state` is more efficient than checking with the string value of title. It is possible for users to intentionally set the milestone title as 'Without a milestone' but not possible for them to set the state to `null`

Echomo-Xinyu replied:
Thanks for the suggestion!

Eclipse-Dominator reviewed (changes requested):
Features works as suggested. Good work!

Eclipse-Dominator reviewed (approved):
LGTM!

kkangs0226 reviewed (changes requested):
Seems like this PR is still based on Angular v8, please resolve conflicts first! Using Angular v9, I am running into errors.

Echomo-Xinyu commented:
> Seems like this PR is still based on Angular v8, please resolve conflicts first! Using Angular v9, I am running into errors.

Fixed!

kkangs0226 reviewed (approved):
LGTM, good work!

chunweii commented:
Before we can merge this PR, please follow the commit message guidelines
